astra/src/main/java/com/slack/astra/bulkIngestApi/BulkIngestS3Producer.java

```diff
@@ -0,0 +1,197 @@
+          compressedData.length,
+          objectKey);
+
+    } catch (Exception e) {
+      LOG.error("Failed to upload to S3", e);
+      throw new RuntimeException("S3 upload failed", e);
+    } finally {
+      uploadTimer.stop(s3UploadTimer);
+    }
+
+    for (Map.Entry<String, List<Trace.Span>> indexDoc : indexDocs.entrySet()) {
+      String index = indexDoc.getKey();
+      List<Trace.Span> spans = indexDoc.getValue();
+      int partition = getPartition(index);
+
+      if (partition < 0) {
+        LOG.warn("index=" + index + " does not have a provisioned dataset associated with it");
+        continue; // Skip this index if no partition is found
+      }
+      // prepare pointer message
+      WalProtos.S3WalPointer pointer =
+          WalProtos.S3WalPointer.newBuilder()
+              .setS3Bucket(walBucket)
+              .setS3Key(objectKey)
+              .setDocCount(spans.size())
+              .setTimestampMs(Instant.now().toEpochMilli())
+              .setCompressionType("gzip")
+              .build();
+
+      byte[] pointerBytes = pointer.toByteArray();
+
+      ProducerRecord<String, byte[]> producerRecord =
+          new ProducerRecord<>(kafkaTopic, partition, index, pointerBytes);
+
+      try {
+        RecordMetadata recordMetadata = this.kafkaProducer.send(producerRecord).get();
+        LOG.debug(
+            "Sent WAL pointer for index {} to Kafka topic {} partition {} offset {}",
+            index,
+            kafkaTopic,
+            recordMetadata.partition(),
+            recordMetadata.offset());
+
+      } catch (Exception e) {
+        LOG.error(
+            "Failed to send WAL pointer for index {} to Kafka - deleting S3 object {}",
+            index,
+            objectKey,
+            e);
+        DeleteObjectRequest deleteRequest =
+            DeleteObjectRequest.builder().bucket(walBucket).key(objectKey).build();
+        s3Client.deleteObject(deleteRequest).join();
+        throw new RuntimeException("Failed to send WAL pointer to Kafka", e);
+      }
+    }
+    // Increment metrics
+    s3UploadCounter.increment();
+    s3SpansUploadedCounter.increment(totalDocs);
+    s3BytesUploadedCounter.increment(compressedData.length);
+
+    return new BulkIngestResponse(totalDocs, 0, "Success");
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    if (s3Client != null) {
+      s3Client.close();
+    }
+    super.shutDown();
+  }
+
+  // generate a key based on the current timestamp and a UUID.
+  private String generateS3ObjectKey() {
+    Instant now = Instant.now();
+    // Create hour based directory structure
+    return String.format(
+        "wal/%d/%02d/%02d/%02d/batch-%d-%s.gz",
+        now.atZone(ZoneOffset.UTC).getYear(),
+        now.atZone(ZoneOffset.UTC).getMonthValue(),
+        now.atZone(ZoneOffset.UTC).getDayOfMonth(),
+        now.atZone(ZoneOffset.UTC).getHour(),
+        now.toEpochMilli(),
+        UUID.randomUUID().toString().substring(0, 8));
```

Comment:
I think it would make sense to include the partition in the object key, since partition maps to an index node that would be downloading the s3 object. It'd make it easier to inspect the s3 contents, or to replay the wal for a given index node.

Reply:
Since we won't be inspecting these contents much, it may be simpler and more efficient to encode the timestamp in millis here. `String ts = Long.toString(System.currentTimeMillis())`.

Also, it may be worth creating a folder per hour and uploading the data in there. That way the folders will be easier to clean up later in S3. This also makes list calls very cheap and fast.